File: src/main/java/de/dc/planetsudo/game/strategy/StrategieNoxus.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package de.dc.planetsudo.game.strategy;

import de.dc.planetsudo.level.levelobjects.AgentInterface;
import de.dc.planetsudo.level.levelobjects.Resource;

/**
 *
 * @author divine
 */
public class StrategieNoxus extends AbstractStrategy {

	public StrategieNoxus() {
	}
	public StrategieNoxus(AgentInterface a) {
		super(a);
	}

	/**
	 * Wie viele Agenten sollen erstellt werde wird hier angegeben.
	 * @return Anzahl der Agenten
	 */
	@Override
	protected int loadAgentCount() {
		return 2;
	}

	@Override
	protected void loadRules() {
		//-------------------------------------------->
		createRule(new Rule(99, "Kollision!") {
			@ Override
			protected boolean constraint() {
				return agent.isCollisionDetected();
			}
			@ Override
			protected void action() {
				agent.turnRandom();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(87, "Warte...") {
			@ Override
			protected boolean constraint() {
				return agent.getFuelInPercent() <= 6 && agent.isAtMothership() && !mothership.hasFuel() && !agent.isUnderAttack() && !mothership.isDamaged();
			}
			@ Override
			protected void action() {
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(86, "Auftanken...") {
			@ Override
			protected boolean constraint() {
				return agent.getFuelInPercent() <= 7 && agent.isAtMothership() && mothership.hasFuel();
			}
			@ Override
			protected void action() {
				agent.orderFuel(100);
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(85, "Wenig Energie...") {
			@ Override
			protected boolean constraint() {
				return agent.getFuelInPercent() <= 7 && agent.hasFuel() && !agent.isUnderAttack();
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(72, "Repariere Mutterschiff...") {
			@ Override
			protected boolean constraint() {
				return mothership.isDamaged() && agent.isAtMothership() && !agent.isUnderAttack();
			}
			@ Override
			protected void action() {
				agent.repairMothership();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(80, "Mutterschiff helfen...") {
			@ Override
			protected boolean constraint() {
				return mothership.isDamaged() && !agent.isAtMothership();
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(75, "Bekämpfe Agent...") {
			@ Override
			protected boolean constraint() {
				return agent.seeAdversaryAgent();
			}
			@ Override
			protected void action() {
				agent.fightWithAdversaryAgent();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(74, "Suche Feind...") {
			@ Override
			protected boolean constraint() {
				return agent.isUnderAttack();
			}
			@ Override
			protected void action() {
				agent.turnRandom();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(61, "Mutterschiff verminen...") {
			@ Override
			protected boolean constraint() {
				return agent.seeAdversaryMothership() && agent.hasMine();
			}
			@ Override
			protected void action() {
				agent.deployMine();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(60, "Bekämpfe Mutterschiff...") {
			@ Override
			protected boolean constraint() {
				return agent.seeAdversaryMothership();
			}
			@ Override
			protected void action() {
				agent.deployMarker();
				agent.fightWithAdversaryMothership();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(35, "Notkanal schließen.") {
			@ Override
			protected boolean constraint() {
				return agent.isSupportOrdered() && agent.getFuelInPercent() >=7;
			}
			@ Override
			protected void action() {
				agent.cancelSupport();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(34, "Gestrandet...") {
			@ Override
			protected boolean constraint() {
				return !agent.hasFuel() && !agent.isAtMothership();
			}
			@ Override
			protected void action() {
				agent.orderSupport();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(31, "Energie Spenden...") {
			@ Override
			protected boolean constraint() {
				return mothership.needSomeoneSupport() && agent.seeLostTeamAgent();
			}
			@ Override
			protected void action() {
				agent.spendTeamAgentFuel(7);
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(30, "Agenten Helfen...") {
			@ Override
			protected boolean constraint() {
				return mothership.needSomeoneSupport();
			}
			@ Override
			protected void action() {
				agent.goToSupportAgent();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(25, "Resource abliefern!") {
			@ Override
			protected boolean constraint() {
				return agent.isAtMothership() && agent.isCarringResource();
			}
			@ Override
			protected void action() {
				agent.deliverResourceToMothership();
				agent.turnAround();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(23, "Berge Resource...") {
			@ Override
			protected boolean constraint() {
				return agent.isCarringResource();
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(22, "Mine!") {
			@ Override
			protected boolean constraint() {
				return (agent.isTouchingResource(Resource.ResourceType.Mine) || agent.isTouchingResource(Resource.ResourceType.ExtremPoint));
			}
			@ Override
			protected void action() {
				agent.go();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(21, "Resource aufheben!") {
			@ Override
			protected boolean constraint() {
				return agent.isTouchingResource() && !agent.isTouchingResource(Resource.ResourceType.Mine) && !agent.isTouchingResource(Resource.ResourceType.ExtremPoint);
			}
			@ Override
			protected void action() {
				agent.pickupResource();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(20, "Resource gesichtet...") {
			@ Override
			protected boolean constraint() {
				return agent.seeResource();
			}
			@ Override
			protected void action() {
				agent.goToResource();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(3, "Zwischentanken...") {
			@ Override
			protected boolean constraint() {
				return agent.isAtMothership() && mothership.hasFuel() && !mothership.isBurning() && !mothership.isDamaged() && !agent.isUnderAttack() && agent.getFuelInPercent()<=95;
			}
			@ Override
			protected void action() {
				agent.orderFuel(100);
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(2, "Marker entfernen.") {
			@ Override
			protected boolean constraint() {
				return agent.seeMarker();
			}
			@ Override
			protected void action() {
				mothership.clearMarker();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(1, "Bewege zu Marker...") {
			@ Override
			protected boolean constraint() {
				return mothership.isMarkerDeployed();
			}
			@ Override
			protected void action() {
				agent.goToMarker();
			}
		});
		//-------------------------------------------->
		//-------------------------------------------->
		createRule(new Rule(0, "Bewegung...") {
			@ Override
			protected boolean constraint() {
				return true;
			}
			@ Override
			protected void action() {
				agent.go();
			}
		});
		//-------------------------------------------->
	}
}

File: src/main/kotlin/org/openbase/planetsudo/game/strategy/StrategieNoxus.kt
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.openbase.planetsudo.game.strategy

import org.openbase.planetsudo.game.SwatTeam
import org.openbase.planetsudo.level.levelobjects.AgentInterface
import org.openbase.planetsudo.level.levelobjects.Resource.ResourceType

/**
 *
 * @author [Divine Threepwood](mailto:divine@openbase.org)
 */
class StrategieNoxus(agent: AgentInterface) : StrategyLevel3(agent) {

    /**
     * Hier wird angegeben wie viele Agenten dem Team zur Verfügung stehen sollen.
     * @return Anzahl der Agenten
     */
    override fun loadAgentCount(): Int {
        return 2
    }

    /**
     * Hier kannst du SwatTeams aus mehreren Agenten bilden.
     * =====================================================
     * Die Agenten werden hierbei über ihre IDs hinzugefügt. Sind beispielsweise 4 Agenten in der Strategie angegeben,
     * so sind diese über die IDs 0 - 3 referenzierbar wobei Agent 0 immer für den Commander steht.
     * Bitte beachte somit, dass die Agenten ID nicht größer als N - 1 sein kann sofern N für die maximale Anzahl von Agenten steht.
     *
     * Die default Gruppen ALL und COMMANDER können anhand dieser Methode nicht modifiziert werden!
     */
    override fun loadSwatTeams() {
        createSwat(SwatTeam.ALPHA, 1)
    }

    override fun loadRules() {
        // -------------------------------------------->
        createRule(
            object : Rule(1000, "Kollision!", SwatTeam.ALL) {
                override fun constraint(): Boolean {
                    return agent.isCollisionDetected
                }

                override fun action() {
                    agent.turnRandom()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(870, "Warte...") {
                override fun constraint(): Boolean {
                    return agent.fuelInPercent <= 6 && agent.isAtMothership && !mothership.hasFuel() && !agent.isUnderAttack && !mothership.isDamaged && !agent.isCarryingResource
                }

                override fun action() {
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(860, "Auftanken...") {
                override fun constraint(): Boolean {
                    return agent.fuelInPercent <= 7 && agent.isAtMothership && mothership.hasFuel()
                }

                override fun action() {
                    agent.orderFuel(100)
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(850, "Wenig Energie...") {
                override fun constraint(): Boolean {
                    return agent.fuelInPercent <= 7 && agent.hasFuel && !agent.isUnderAttack
                }

                override fun action() {
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(720, "Repariere Mutterschiff...") {
                override fun constraint(): Boolean {
                    return mothership.isDamaged && agent.isAtMothership && !agent.isUnderAttack
                }

                override fun action() {
                    agent.repairMothership()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(800, "Mutterschiff helfen...") {
                override fun constraint(): Boolean {
                    return mothership.isDamaged && !agent.isAtMothership
                }

                override fun action() {
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(750, "Bekämpfe Agent...") {
                override fun constraint(): Boolean {
                    return agent.seeAdversaryAgent
                }

                override fun action() {
                    agent.fightWithAdversaryAgent()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(740, "Suche Feind...") {
                override fun constraint(): Boolean {
                    return agent.isUnderAttack
                }

                override fun action() {
                    agent.turnRandom()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(610, "Mutterschiff verminen...") {
                override fun constraint(): Boolean {
                    return agent.seeAdversaryMothership && agent.hasMine
                }

                override fun action() {
                    agent.deployMine()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(600, "Bekämpfe Mutterschiff...") {
                override fun constraint(): Boolean {
                    return agent.seeAdversaryMothership
                }

                override fun action() {
                    agent.deployMarker()
                    agent.fightWithAdversaryMothership()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(350, "Notkanal schließen.") {
                override fun constraint(): Boolean {
                    return agent.isSupportOrdered && agent.fuelInPercent >= 7
                }

                override fun action() {
                    agent.cancelSupport()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(340, "Gestrandet...") {
                override fun constraint(): Boolean {
                    return !agent.hasFuel && !agent.isAtMothership
                }

                override fun action() {
                    agent.orderSupport()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(310, "Energie Spenden...") {
                override fun constraint(): Boolean {
                    return mothership.needSomeoneSupport() && agent.seeLostTeamAgent()
                }

                override fun action() {
                    agent.spendTeamAgentFuel(7)
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(300, "Agenten Helfen...") {
                override fun constraint(): Boolean {
                    return mothership.needSomeoneSupport()
                }

                override fun action() {
                    agent.goToSupportAgent()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(250, "Resource abliefern!") {
                override fun constraint(): Boolean {
                    return agent.isAtMothership && agent.isCarryingResource
                }

                override fun action() {
                    agent.deliverResourceToMothership()
                    agent.turnAround()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(230, "Berge Resource...") {
                override fun constraint(): Boolean {
                    return agent.isCarryingResource
                }

                override fun action() {
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(220, "Mine!") {
                override fun constraint(): Boolean {
                    return (agent.isTouchingResource(ResourceType.Mine) || agent.isTouchingResource(ResourceType.ExtremPoint))
                }

                override fun action() {
                    agent.go()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(210, "Resource aufheben!") {
                override fun constraint(): Boolean {
                    return agent.isTouchingResource && !agent.isTouchingResource(ResourceType.Mine) && !agent.isTouchingResource(
                        ResourceType.ExtremPoint,
                    )
                }

                override fun action() {
                    agent.pickupResource()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(200, "Resource gesichtet...") {
                override fun constraint(): Boolean {
                    return agent.seeResource
                }

                override fun action() {
                    agent.goToResource()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(30, "Zwischentanken...") {
                override fun constraint(): Boolean {
                    return agent.isAtMothership && mothership.hasFuel() && !mothership.isBurning && !mothership.isDamaged && !agent.isUnderAttack && agent.fuelInPercent <= 95
                }

                override fun action() {
                    agent.orderFuel(100)
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(20, "Marker entfernen.") {
                override fun constraint(): Boolean {
                    return agent.seeMarker()
                }

                override fun action() {
                    mothership.clearMarker()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(10, "Bewege zu Marker...") {
                override fun constraint(): Boolean {
                    return mothership.isMarkerDeployed
                }

                override fun action() {
                    agent.goToMarker()
                }
            },
        )
        // -------------------------------------------->
        // -------------------------------------------->
        createRule(
            object : Rule(0, "Bewegung...") {
                override fun constraint(): Boolean {
                    return true
                }

                override fun action() {
                    agent.go()
                }
            },
        )
        // -------------------------------------------->
    }
}
